### ISSUE TITLE
Exception handling when using OCI_CHARSET_WIDE on MAC

### ISSUE THREAD

**Issue by AidyCC** [vrogier/ocilib#250]
I'm using ocilib on MAC OSX and want to build the OCI_CHARSET_WIDE version.

The library builds fine and I've got ocilib.demo running in XCode.

However, the exception handling does not work.  I believe it's because the Exception::what() returns char* (because that's that the base class does).

I get a compilation error in Execption.hpp:-

<img width="1004" alt="Screenshot 2020-11-10 at 18 07 26" src="https://user-images.githubusercontent.com/74263551/98713561-ae7a4700-237f-11eb-8ce5-e0ac7378ba52.png">

The compilation error can be resovled by doing this:-

<img width="470" alt="Screenshot 2020-11-10 at 18 08 54" src="https://user-images.githubusercontent.com/74263551/98713791-f39e7900-237f-11eb-91c6-018f6e0b94ac.png">

But this does not help in reporing the error in the "catch" handler.

I assume I was able to build the library with OCI_CHARSET_WIDE because these methods are inline so only get compiled when they are used.

What is the correct way to handle excpetions when using OCI_CHARSET_WIDE based builds?  Am I missing something?

-AidyCC

**Comment by AidyCC** [comment]
Seems this is issue is related to a change I had to make to Stringutils.c :-

<img width="763" alt="Screenshot 2020-11-11 at 13 19 41" src="https://user-images.githubusercontent.com/74263551/98816579-cb1a8b80-2420-11eb-9219-b35f9492ed26.png">

Is there an alternative to vsnwprintf ?

**Comment by vrogier** [comment]
Hi,

I will create a v4.7.3 branch today with the fix for the C+X Ocilib::Exception::what() :)
about the vsnwprintf not available on mac( was not aware of it), why not something like :


```c
int OcilibStringFormat
(
    otext       * str,
    int           size,
    const otext * format,
    ...
)
{
    va_list args;

    va_start(args, format);

#ifdef OCI_CHARSET_ANSI

    #ifdef __APPLE__
        const int n = (int)vsprintf(str, format, args);
    #else
        const int n = (int)vsnprintf(str, (size_t)size, format, args);
    #endif

#else

    #ifdef __APPLE__
        const int n = (int)vswprintf(str, format, args);
    #else
        const int n = (int)vsnwprintf(str, (size_t)size, format, args);
    #endif

#endif

    va_end(args);

    return n;
}
```


Vincent

**Comment by AidyCC** [comment]
Thanks Vincent.  I'll give that a go.  That would be a lot more elegant than my current solution which I was just about to post.

-AidyCC

**Comment by AidyCC** [comment]
I changed the Exception class for WIDE builds only as follows:-

<img width="439" alt="Screenshot 2020-11-12 at 11 15 14" src="https://user-images.githubusercontent.com/74263551/98935987-3d03db00-24dc-11eb-8c73-6930f1006abd.png">

<img width="428" alt="Screenshot 2020-11-12 at 11 15 47" src="https://user-images.githubusercontent.com/74263551/98936033-4d1bba80-24dc-11eb-939c-b075a197dc24.png">

<img width="427" alt="Screenshot 2020-11-12 at 11 16 24" src="https://user-images.githubusercontent.com/74263551/98936055-5573f580-24dc-11eb-8dcb-cd8eced9b690.png">

I then changed OcilibStringFormat as follows based on your suggestion:-

<img width="570" alt="Screenshot 2020-11-12 at 11 45 23" src="https://user-images.githubusercontent.com/74263551/98936211-92d88300-24dc-11eb-97e8-901283b5f7b6.png">

When running the ocilib_demo.cpp supplying a wrong password to generate an exception I see this:-

<img width="427" alt="Screenshot 2020-11-12 at 11 47 29" src="https://user-images.githubusercontent.com/74263551/98936451-e3e87700-24dc-11eb-8a02-f4a6232d9300.png">

So there is an issue with encoding %s placeholders.

Modifying the WIDE branch of OcilibStringFormat as follows:-

<img width="606" alt="Screenshot 2020-11-12 at 11 52 55" src="https://user-images.githubusercontent.com/74263551/98937015-b8b25780-24dd-11eb-9e6a-845512a85756.png">

Results in this:-

<img width="415" alt="Screenshot 2020-11-12 at 11 54 46" src="https://user-images.githubusercontent.com/74263551/98937119-e13a5180-24dd-11eb-83f0-7502aad52741.png">

So there is an issue with the string returned by OCIErrorGet in Wide mode..

Modifying OcilibExceptionOCI as follows:-

<img width="677" alt="Screenshot 2020-11-12 at 11 57 33" src="https://user-images.githubusercontent.com/74263551/98937416-468e4280-24de-11eb-87f8-ce56df9eb203.png">

Results in this:-

<img width="738" alt="Screenshot 2020-11-12 at 11 58 15" src="https://user-images.githubusercontent.com/74263551/98937502-6160b700-24de-11eb-9447-0af7066f2980.png">

Which is the correct output.

Like I said, not an elegant solution.  I look forward to seeing your update.

-AidyCC

**Comment by vrogier** [comment]
Hi,

In the meanwhile, I have create a develop-v4.7.3 branch and commit a fix for Exception::what() 

Regards,

Vincent

**Comment by vrogier** [comment]
About the formatting issue, i have pushed another commit when doing internal formatting in OCI_CHARSET_WIDE on linux/unix platforms (302a523a)

**Comment by vrogier** [comment]
Can you let me kwown if all issues are resolved ?

thanks

Vincent

**Comment by AidyCC** [comment]
Ok, taking a look now.

**Comment by AidyCC** [comment]
Not quite...  Slight modification needed in Exception::GetMessage() const to move the `return message` outside of the #endif:-

<img width="679" alt="Screenshot 2020-11-12 at 14 47 26" src="https://user-images.githubusercontent.com/74263551/98954753-06d35500-24f6-11eb-9140-e5f87e3c250c.png">

Output of a OCI_CHARSET_ANSI build (OK) :-

<img width="646" alt="Screenshot 2020-11-12 at 14 43 28" src="https://user-images.githubusercontent.com/74263551/98954315-93c9de80-24f5-11eb-96b8-94c7bf17cd79.png">

Output of a OCI_CHARSET_WIDE build (suspected OCIErrorGet encoding issue):-

<img width="455" alt="Screenshot 2020-11-12 at 14 43 51" src="https://user-images.githubusercontent.com/74263551/98954339-9a585600-24f5-11eb-9314-3c8ab6a639d2.png">

**Comment by vrogier** [comment]
Hi,

I got why... and pushed another commit. I will make test under Linux later today (quite busy now).
Can you have try in the meanwhile ?

Vincent

**Comment by AidyCC** [comment]
Sure.  Thanks for the quick turn around.

**Comment by AidyCC** [comment]
Yep!  Those latest changes did the trick.

Thanks,

-AidyCC

**Comment by vrogier** [comment]
good 👍 

Let me know if you find other issues related to OCI_CHARSET_WIDE on macos.

I'll create an issue for the error message retrieval and add tests case for both issues and will make release by the end of the week I guess

Best regards,

Vincent
